apps/service_providers/tracing/service.py

```diff
@@ -0,0 +1,204 @@
+        self.trace_id: UUID | None = None
+        self.session_id: str | None = None
+        self.user_id: str | None = None
+
+    @classmethod
+    def empty(cls) -> Self:
+        return cls([])
+
+    @classmethod
+    def create_for_experiment(cls, experiment) -> Self:
+        tracers = []
+        if experiment and experiment.trace_provider:
+            try:
+                tracers.append(experiment.trace_provider.get_service())
+            except Exception as e:  # noqa: BLE001
+                logger.error(f"Error setting up trace service: {e}")
+
+        return cls(tracers)
+
+    @contextmanager
+    def trace(self, trace_name: str, session_id: str, user_id: str):
+        self.session_id = session_id
+        self.trace_name = trace_name
+        self.user_id = user_id
+        self.trace_id = uuid.uuid4()
+
+        try:
+            self._begin_traces()
+            self.activated = True
+            yield self
+        finally:
+            self._end_traces()
+
+    def _begin_traces(self):
+        for tracer in self._tracers:
+            try:
+                tracer.begin_trace(self.trace_name, self.trace_id, self.session_id, self.user_id)
+            except Exception:  # noqa BLE001
+                logger.error("Error initializing tracer %s", tracer.__class__.__name__, exc_info=True)
+
+    def _end_traces(self):
+        for tracer in self._active_tracers:
+            try:
+                tracer.end_trace()
+            except Exception:  # noqa BLE001
+                logger.error("Error ending tracer %s", tracer.__class__.__name__, exc_info=True)
+        self._reset_io()
+
+    @contextmanager
+    def span(
+        self,
+        span_id: str,
+        span_name: str,
+        inputs: dict[str, Any],
+        metadata: dict[str, Any] | None = None,
+    ):
+        if not self.activated:
+            yield self
+            return
+
+        self._start_span(
+            span_id,
+            span_name,
+            inputs,
+            metadata,
+        )
+        try:
+            yield self
+        except Exception as e:
+            self._end_span(span_id, span_name, e)
+            raise
+        else:
+            self._end_span(span_id, span_name)
+
+    def set_outputs(
+        self,
+        span_id: str,
+        outputs: dict[str, Any],
+        output_metadata: dict[str, Any] | None = None,
+    ) -> None:
+        self.outputs[span_id] |= outputs or {}
+        self.outputs_metadata[span_id] |= output_metadata or {}
+
+    def get_langchain_callbacks(self) -> list["BaseCallbackHandler"]:
+        if not self.activated:
+            return []
+
+        callbacks = []
+        for tracer in self._active_tracers:
+            callback = tracer.get_langchain_callback()
+            if callback:
+                callbacks.append(wrap_callback(callback))
+
+        return callbacks
+
+    def get_langchain_config(self, *, callbacks: list = None, configurable: dict = None) -> RunnableConfig:
```

Comment:
Just curious: When we implement our own tracers, we'd have to implement `get_langchain_callback`. How do you envision this would be used by ocs tracers?

Reply:
It would be similar to how we currently use the `PipelineLoggingCallbackHandler`